Failure in InstructorFeedbackResultsPageUiTest.testPanelsCollapseExpand

java.lang.AssertionError: 
	at org.testng.AssertJUnit.fail(AssertJUnit.java:59)
	at org.testng.AssertJUnit.assertTrue(AssertJUnit.java:24)
	at org.testng.AssertJUnit.assertFalse(AssertJUnit.java:41)
	at org.testng.AssertJUnit.assertFalse(AssertJUnit.java:49)
	at teammates.test.pageobjects.InstructorFeedbackResultsPage.verifyParticipantPanelIsCollapsed(InstructorFeedbackResultsPage.java:419)
	at teammates.test.cases.ui.browsertests.InstructorFeedbackResultsPageUiTest.testPanelsCollapseExpand(InstructorFeedbackResultsPageUiTest.java:489)
	at teammates.test.cases.ui.browsertests.InstructorFeedbackResultsPageUiTest.testFrontEndActions(InstructorFeedbackResultsPageUiTest.java:66)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:84)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:714)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:901)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1231)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:127)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:111)
	at org.testng.TestRunner.privateRun(TestRunner.java:767)
	at org.testng.TestRunner.run(TestRunner.java:617)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:334)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:329)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:291)
	at org.testng.SuiteRunner.run(SuiteRunner.java:240)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
	at org.testng.TestNG.run(TestNG.java:1057)
	at org.testng.remote.RemoteTestNG.run(RemoteTestNG.java:111)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:204)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:175)

Getting a similar error (on dev server) from a different part of the test:

java.lang.AssertionError: 
	at org.testng.AssertJUnit.fail(AssertJUnit.java:59)
	at org.testng.AssertJUnit.assertTrue(AssertJUnit.java:24)
	at org.testng.AssertJUnit.assertFalse(AssertJUnit.java:41)
	at org.testng.AssertJUnit.assertFalse(AssertJUnit.java:49)
	at teammates.test.pageobjects.InstructorFeedbackResultsPage.verifyParticipantPanelIsCollapsed(InstructorFeedbackResultsPage.java:462)
	at teammates.test.cases.ui.browsertests.InstructorFeedbackResultsPageUiTest.testPanelsCollapseExpand(InstructorFeedbackResultsPageUiTest.java:497)
	at teammates.test.cases.ui.browsertests.InstructorFeedbackResultsPageUiTest.testFrontEndActions(InstructorFeedbackResultsPageUiTest.java:66)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:84)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:714)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:901)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1231)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:127)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:111)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:744)

--- a/src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackResultsPageUiTest.java
+++ b/src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackResultsPageUiTest.java
@@ -491,10 +491,9 @@ public class InstructorFeedbackResultsPageUiTest extends BaseUiTestCase {
         resultsPage.displayByGiverRecipientQuestion();
         assertEquals("Collapse Students", resultsPage.instructorPanelCollapseStudentsButton.getText());
         resultsPage.clickInstructorPanelCollapseStudentsButton();
+        resultsPage.waitForInstructorPanelStudentPanelsToCollapse();
         assertEquals("Expand Students", resultsPage.instructorPanelCollapseStudentsButton.getText());
-        
-        int numOfPanels = resultsPage.getNumOfPanelsInInstructorPanel();
-        resultsPage.verifyParticipantPanelIsCollapsed(4, (numOfPanels * 50));
+        resultsPage.verifySpecifiedPanelIdsAreCollapsed(new int[] { 3, 4, 5 });
 
         resultsPage.clickGroupByTeam();
         resultsPage.displayByQuestion();

--- a/src/test/java/teammates/test/pageobjects/AppPage.java
+++ b/src/test/java/teammates/test/pageobjects/AppPage.java
@@ -27,11 +27,13 @@ import org.openqa.selenium.JavascriptExecutor;
 import org.openqa.selenium.Keys;
 import org.openqa.selenium.NoAlertPresentException;
 import org.openqa.selenium.NoSuchElementException;
+import org.openqa.selenium.WebDriver;
 import org.openqa.selenium.WebElement;
 import org.openqa.selenium.remote.RemoteWebElement;
 import org.openqa.selenium.remote.UselessFileDetector;
 import org.openqa.selenium.support.FindBy;
 import org.openqa.selenium.support.PageFactory;
+import org.openqa.selenium.support.ui.ExpectedCondition;
 import org.openqa.selenium.support.ui.ExpectedConditions;
 import org.openqa.selenium.support.ui.Select;
 import org.openqa.selenium.support.ui.WebDriverWait;
@@ -215,6 +217,40 @@ public abstract class AppPage {
         wait.until(ExpectedConditions.invisibilityOfElementLocated(by));
     }
     
+    /**
+     * Waits for a list of elements to be invisible or not present, or timeout.
+     */
+    public void waitForElementsToDisappear(List<WebElement> elements) {
+        WebDriverWait wait = new WebDriverWait(browser.driver, TestProperties.inst().TEST_TIMEOUT);
+        wait.until(invisibilityOfAllElements(elements));
+    }
+    
+    /**
+     * Code adapted from SeleniumHQ's GitHub page.
+     * TODO to be removed when Selenium is upgraded to the version supporting this method.
+     *
+     * An expectation for checking all elements from given list to be invisible
+     *
+     * @param elements used to check their invisibility
+     * @return Boolean true when all elements are not visible anymore
+     */
+    private ExpectedCondition<Boolean> invisibilityOfAllElements(final List<WebElement> elements) {
+        return new ExpectedCondition<Boolean>() {
+            @Override
+            public Boolean apply(WebDriver webDriver) {
+                for (WebElement element : elements) {
+                    try {
+                        if (element.isDisplayed()) {
+                            return false;
+                        }
+                    } catch (Exception e) {
+                    }
+                }
+                return true;
+            }
+        };
+    }
+
     /**
      * Waits for the element to appear in the page, up to the timeout specified.
      */

--- a/src/test/java/teammates/test/pageobjects/AppPage.java
+++ b/src/test/java/teammates/test/pageobjects/AppPage.java
@@ -248,6 +248,11 @@ public abstract class AppPage {
                 }
                 return true;
             }
+
+            @Override
+            public String toString() {
+                return "invisibility of all elements " + elements;
+            }
         };
     }
 

--- a/src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java
+++ b/src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java
@@ -42,9 +42,6 @@ public class InstructorFeedbackResultsPage extends AppPage {
     @FindBy(id = "show-stats-checkbox")
     public WebElement showStatsCheckbox;
     
-    @FindBy(id = "panelBodyCollapse-1")
-    public WebElement instructorPanelBody;
-
     public InstructorFeedbackResultsPage(Browser browser) {
         super(browser);
     }
@@ -453,19 +450,16 @@ public class InstructorFeedbackResultsPage extends AppPage {
         instructorPanelCollapseStudentsButton.click();
     }
 
-    public void verifyParticipantPanelIsCollapsed(int id, int timeToWait) {
-        WebElement panel = browser.driver.findElement(By.id("panelBodyCollapse-" + id));
-
-        // Need to wait for the total duration according to the number of collapse/expand intervals 
-        // between panels before checking final state of the panel
-        ThreadHelper.waitFor(timeToWait);
-        assertFalse(panel.isDisplayed());
+    public void waitForInstructorPanelStudentPanelsToCollapse() {
+        List<WebElement> studentPanels = browser.driver.findElements(By.cssSelector("#panelBodyCollapse-2 .panel-collapse"));
+        waitForElementsToDisappear(studentPanels);
     }
 
-    public int getNumOfPanelsInInstructorPanel() {
-        List<WebElement> participantPanels = instructorPanelBody
-                                                 .findElements(By.xpath(".//div[contains(@class, 'panel-collapse')]"));
-        return participantPanels.size();
+    public void verifySpecifiedPanelIdsAreCollapsed(int[] ids) {
+        for (int id : ids) {
+            WebElement panel = browser.driver.findElement(By.id("panelBodyCollapse-" + id));
+            assertFalse(panel.isDisplayed());
+        }
     }
 
     public boolean isSectionPanelExist(String section) {

--- a/src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java
+++ b/src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java
@@ -251,7 +251,7 @@ public class InstructorFeedbackResultsPage extends AppPage {
     public void waitForPanelsToCollapse() {
         By panelCollapseSelector = By.cssSelector("div[id^='panelBodyCollapse-']");
         
-        waitForElementToDisappear(panelCollapseSelector);
+        waitForElementsToDisappear(browser.driver.findElements(panelCollapseSelector));
     }
     
     /**

